sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -798,10 +830,35 @@ class CodegenContext {
            |  ${makeSplitFunction(body)}
            |}
          """.stripMargin
-        addNewFunction(name, code)
+        addNewFunctionInternal(name, code, inlineToOuterClass = false)
       }
 
-      foldFunctions(functions.map(name => s"$name(${arguments.map(_._2).mkString(", ")})"))
+      val outerClassFunctions = functions
+        .filter(_.subclassName.isEmpty)
+        .map(_.functionName)
+
+      val innerClassFunctions = functions
+        .filter(_.subclassName.isDefined)
+        .foldLeft(Map.empty[(String, String), Seq[String]]) { case (acc, f) =>
+          val key = (f.subclassName.get, f.subclassInstance.get)
+          acc.updated(key, acc.getOrElse(key, Seq.empty[String]) ++ Seq(f.functionName))
+        }
+        .map { case ((subclassName, subclassInstance), subclassFunctions) =>
+          // Adding a new function to each subclass which contains
+          // the invocation of all the ones which have been added to
+          // that subclass
+          val code = s"""
+              |private $returnType $func($argString) {
+              |  ${makeSplitFunction(foldFunctions(subclassFunctions.map(name =>
+                    s"$name(${arguments.map(_._2).mkString(", ")})")))}
+              |}
+            """.stripMargin
+          addNewFunctionToClass(func, code, subclassName)
```

Comment:
The class to add codes into can be over the threshold too.

Reply:
So shall we call `addNewFunction` to add the caller method into an inner class and call those split functions in that?

Reply:
no, this is made intentionally because if we add it to the {{NestedClass}} where these methods are added: they are not available in other nested classes...

Reply:
Can't we call them with instance?

Reply:
we might, but I think this would not be the best option, since in this way we are "reusing" some ConstantPool entries (like `Utf8` for the name of the function) and it keeps everything about those things in the same class.
Moreover, foreach instance referenced from other classes, like it would be in that case, we are adding new entries to the Constant Pool of that class.
Thus this is the option which minimizes the number of entries added to the Constant Pools.

Reply:
Because `addNewFunction` is not only used by `splitExpressions`, this threshold change may increase the number of nested sub-classes and then of course the instances. So I think it is also bad for the Constant Pool limit?

Reply:
Worse, I guess the functions added by `addNewFunction` are called at the outerclass at most, so basically this way is going to increase the Constant Pool entries in the outerclass.

I agreed that in the suggested way, we increase Constant Pool entries in the inner classes, but I think inner classes can be more free from this issue?

This is just my guess. So please let me know if I reason it incorrectly.

Reply:
yes,  for each new `NestedClassX` we have a new `nestedClassInstanceX` attribute to the outer class. Every attribute brings two entries in the constant pool. Currently, I have never been able to generate a class with more than 1-2 nested classes. Since we are roughly setting this value to half of the previous one, we can say that we are going to have 2-4 nested classes, which means we are adding at most 10 entries in the constant pool of the outer class: I think this is not a big deal.

Reply:
sorry, I didn't get completely you last comments, might you elaborate it a bit more? Thanks.

Reply:
Oh. I meant the increased inner classes/instances are added into the outerclass's Constant Pool entries. Outerclass should be more likely to be filled by various variables/methods...than inner classes. So I think to add into Constant Pool of inner classes should be less serious issue than outerclass.

Reply:
It is also convincing that nested classes should be less. So maybe this is not a serious issue.

Reply:
yes, but we are likely to recreate the same issue the PR is trying to solve in the last `NestedClass`. The problem is that a method call, despite it is few bytes of code, introduce multiple entries in the constant pool. If we use `addNewFunction` here, we will end up creating all the wrapper methods in the last `NestedClass`, which will have the same issues as the outer class, becoming a new bottelneck.
Instead, in this way, we are splitting the method call among the several nested classes and each of them has its owns, thus it is reusing some constant pool entries.
As you pointed out, nested classes are less critical than the outer one, because the attributes are declared in the outer class and currently all the many small methods are called there. If we put `addNewFunction` here, instead, we are making critical the last nested function because we are adding there all the many small function calls.

Reply:
Sounds reasonable to me.